# Coding Exercises (Part 2)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas: Basics

### Exercise 2: Indexing and Slicing

7. -__Import Pandas__ (as pd)<br>
-__Import__ the Listings Dataset (__listings.csv__) and save the DataFrame in the variable __listings__ and<br>
-__Inspect__ the __first 5 Rows__!

In [1]:
import pandas as pd

In [2]:
df=pd.read_csv("listings.csv")

In [3]:
df.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

8. __Select__ the column __"Name"__ with the __attribute__ notation! What is the __company name__ with __index label 4__?

In [4]:
df.Name[4]

'American Airlines Group, Inc.'

The company name is ... American Airlines Group, Inc.!

9. Apply also the __second alternative__ to select the column __"Name"__!

In [5]:
df["Name"]

0          Agilent Technologies, Inc.
1                   Alcoa Corporation
2                         Altaba Inc.
3                  AAC Holdings, Inc.
4       American Airlines Group, Inc.
                    ...              
6847                      Zumiez Inc.
6848                      Zuora, Inc.
6849                   Zymeworks Inc.
6850    Zynerba Pharmaceuticals, Inc.
6851                      Zynex, Inc.
Name: Name, Length: 6852, dtype: object

10. __Select__ the listing/row at __index position 100__ with the __iloc__ operator (all columns)! What is the IPO_Year?

In [8]:
df.iloc[100,-3]

2014.0

The IPO_Year of this listing is ... 2014!

11. __Select__ the listing/row at __index position 2000__ with the __iloc__ operator (and only the column __"Name"__)! <br>
What is the __company name__ of this listing?

In [9]:
df.iloc[2000,2]

'Eagle Bancorp, Inc.'

It´s the ... Eagle Bancorp, Inc.!

12. __Select__ the __last 10 rows__ and the columns __Name, Last_Price, Sector__ and __Industry__ with the __iloc__ operator! __Fill in the Gaps__! <br>
What is the __Sector__ of the __Zymeworks Inc.__?

In [14]:
df[df.Name=='Zymeworks Inc.'].Sector

6849    Health Care
Name: Sector, dtype: object

The Sector is ... Healthcare!

13. __Import__ the __Listings Dataset__ one more time (save in the variable __listings__) and set the column __"Symbol"__ as the __index__! <br>
__Fill in the Gaps__!

In [17]:
listings = pd.read_csv("listings.csv", index_col = "Symbol")

In [18]:
#run the cell and inspect!
listings.head()

Exchange                           Name  Last_Price    Market_Cap  \
Symbol                                                                     
A          NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
AA         NYSE              Alcoa Corporation       29.15  5.407810e+09   
AABA     NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
AAC        NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
AAL      NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

        ADR TSO  IPO_Year            Sector  \
Symbol                                        
A           NaN    1999.0     Capital Goods   
AA          NaN    2016.0  Basic Industries   
AABA        NaN       NaN        Technology   
AAC         NaN    2014.0       Health Care   
AAL         NaN       NaN    Transportation   

                                                Industry  
Symbol                                                    
A       Biotechnology: Laboratory Analytical Instruments  
AA                                              Aluminum  
AABA                                        EDP Services  
AAC                                 Medical Specialities  
AAL                        Air Freight/Delivery Services

In [19]:
#run the cell and inspect!
listings.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6852 entries, A to ZYXI
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Exchange    6852 non-null   object 
 1   Name        6852 non-null   object 
 2   Last_Price  6745 non-null   float64
 3   Market_Cap  5954 non-null   float64
 4   ADR TSO     140 non-null    float64
 5   IPO_Year    3105 non-null   float64
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(4), object(4)
memory usage: 481.8+ KB


14. __Inspect__ the new Index of the DataFrame by calling the __index__ attribute!

In [23]:
listings.index

Index(['A', 'AA', 'AABA', 'AAC', 'AAL', 'AAMC', 'AAME', 'AAN', 'AAOI', 'AAON',
       ...
       'ZSAN', 'ZTEST', 'ZTO', 'ZTR', 'ZTS', 'ZUMZ', 'ZUO', 'ZYME', 'ZYNE',
       'ZYXI'],
      dtype='object', name='Symbol', length=6852)

15. __Select__ the __row__ with the index label __"AAPL"__ (Apple) (use the __loc__ operator)! What is the __Last_Price__?

In [24]:
listings.loc["AAPL" ,"Last_Price"]

200.62

The Last_Price is ... 200.62!

16. __Select__ the __row__ with the index label __"FB"__ (Facebook) and only the columns __"IPO_Year"__ and __"Industry"__! <br>
What´s the __Industry__?

In [25]:
listings.loc["FB",["IPO_Year","Industry"]]

IPO_Year                                             2012.0
Industry    Computer Software: Programming, Data Processing
Name: FB, dtype: object

The Industry is ... Computer Software: Programming, Data Processing!

17. __Select__ the columns __"Name"__ and __"Last_Price"__ with the __loc__ operator (all rows)!

In [26]:
listings.loc[:,["Name","Last_Price"]]

Name  Last_Price
Symbol                                           
A          Agilent Technologies, Inc.       81.68
AA                  Alcoa Corporation       29.15
AABA                      Altaba Inc.       75.39
AAC                AAC Holdings, Inc.        2.16
AAL     American Airlines Group, Inc.       34.02
...                               ...         ...
ZUMZ                      Zumiez Inc.       26.72
ZUO                       Zuora, Inc.       19.79
ZYME                   Zymeworks Inc.       15.74
ZYNE    Zynerba Pharmaceuticals, Inc.        7.85
ZYXI                      Zynex, Inc.        5.02

[6852 rows x 2 columns]

18. __Select__ the last three columns __"IPO_Year"__, __"Sector"__ and __"Industry"__ for __"MSFT"__ (Microsoft) and __"DIS"__ (Disney) with the __loc__ operator. Don´t use "Sector" or "Industry" in your code! What is Disney´s __Industry__?

In [28]:
listings.loc[['MSFT','DIS'],"IPO_Year":]

IPO_Year             Sector                                 Industry
Symbol                                                                      
MSFT      1986.0         Technology  Computer Software: Prepackaged Software
DIS          NaN  Consumer Services                      Television Services

Disney´s Industry is ... Television Services

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Hints (Spoiler!)

7. pandas as pd, pd.read_csv("filename"), head()

8. listings.---

9. listings["---"]

10. listings.iloc[---]

11. listings.iloc[row index position, column index position]; "Name" column is at index position ... 2

12. The last 10 rows = from index position -10 till the end (:); Count the column index positions of the Columns


13. parameter index_col

14. listings.---

15. listings.loc["row label"]

16. listings.loc[---, [---, ---]]

17.  listings.loc[ : , [---, ---]]

18. listings.loc[[---, ---], ---:]